File: orchestrator-node/src/agents/publishAgent.ts
/**
 * Publishing Agent - Posts content to social media platforms.
 */

import * as crypto from 'crypto';
import { BaseAgent } from './baseAgent';
import { Config } from '../config';
import { SocialPost, ImageSet, PublishResult, PublishStatus, Platform } from '../models';
import { RateLimiter, retryWithBackoff, sleep } from '../utils';

export class PublishAgent extends BaseAgent {
  private config: Config;
  private maxRetries = 3;
  private manualReviewQueue: PublishResult[] = [];

  constructor(config: Config, rateLimiter?: RateLimiter) {
    super('publish_agent', rateLimiter);
    this.config = config;
  }

  async execute(
    posts: SocialPost[],
    imageSets: ImageSet[],
    dryRun = false
  ): Promise<PublishResult[]> {
    this.logger.info(`Publishing ${posts.length} posts (dryRun=${dryRun})`);

    // Create image lookup
    const imagesByPost = new Map<string, ImageSet>();
    for (const imageSet of imageSets) {
      imagesByPost.set(imageSet.postId, imageSet);
    }

    const results: PublishResult[] = [];

    for (const post of posts) {
      const images = imagesByPost.get(post.postId);

      if (dryRun) {
        results.push(this.simulatePublish(post, images));
      } else {
        const result = await this.publishPost(post, images);
        results.push(result);

        // Stagger posts to avoid spam appearance
        await sleep(2000);
      }
    }

    // Log summary
    const successCount = results.filter((r) => r.status === PublishStatus.PUBLISHED).length;
    const failedCount = results.filter((r) => r.status === PublishStatus.FAILED).length;
    const reviewCount = results.filter((r) => r.status === PublishStatus.PENDING_REVIEW).length;

    this.logger.info(
      `Publishing complete: ${successCount} success, ${failedCount} failed, ${reviewCount} pending review`
    );

    return results;
  }

  private async publishPost(post: SocialPost, images?: ImageSet): Promise<PublishResult> {
    const platformHandlers: Record<Platform, (p: SocialPost, i?: ImageSet) => Promise<PublishResult>> = {
      [Platform.TWITTER]: (p, i) => this.publishTwitter(p, i),
      [Platform.LINKEDIN]: (p, i) => this.publishLinkedIn(p, i),
      [Platform.INSTAGRAM]: (p, i) => this.publishInstagram(p, i),
      [Platform.FACEBOOK]: (p, i) => this.publishFacebook(p, i),
      [Platform.TIKTOK]: (p, i) => this.publishTikTok(p, i),
    };

    const handler = platformHandlers[post.platform];
    if (!handler) {
      return {
        postId: post.postId,
        platform: post.platform,
        status: PublishStatus.FAILED,
        errorMessage: `Unsupported platform: ${post.platform}`,
        retryCount: 0,
      };
    }

    let lastError: string | undefined;

    for (let attempt = 0; attempt < this.maxRetries; attempt++) {
      try {
        const result = await handler(post, images);
        if (result.status === PublishStatus.PUBLISHED) {
          return result;
        }
        lastError = result.errorMessage;
      } catch (e) {
        lastError = e instanceof Error ? e.message : String(e);
        this.logger.warn(`Publish attempt ${attempt + 1} failed for ${post.postId}: ${lastError}`);

        if (attempt < this.maxRetries - 1) {
          await sleep(5000 * (attempt + 1));
        }
      }
    }

    // All retries failed - queue for manual review
    const result: PublishResult = {
      postId: post.postId,
      platform: post.platform,
      status: PublishStatus.PENDING_REVIEW,
      errorMessage: `Max retries exceeded: ${lastError}`,
      retryCount: this.maxRetries,
    };
    this.manualReviewQueue.push(result);
    return result;
  }

  private async publishTwitter(post: SocialPost, _images?: ImageSet): Promise<PublishResult> {
    if (!this.config.twitterAccessToken) {
      return {
        postId: post.postId,
        platform: Platform.TWITTER,
        status: PublishStatus.FAILED,
        errorMessage: 'Twitter credentials not configured',
        retryCount: 0,
      };
    }

    if (!(await this.rateLimiter.acquire('twitter_post'))) {
      return {
        postId: post.postId,
        platform: Platform.TWITTER,
        status: PublishStatus.FAILED,
        errorMessage: 'Twitter rate limited',
        retryCount: 0,
      };
    }

    return retryWithBackoff(
      async () => {
        const tweetData: { text: string; media?: { media_ids: string[] } } = {
          text: post.refinedContent ?? post.content,
        };

        const url = 'https://api.twitter.com/2/tweets';
        const authHeader = this.buildOAuth1Header('POST', url);

        const response = await fetch(url, {
          method: 'POST',
          headers: {
            Authorization: authHeader,
            'Content-Type': 'application/json',
          },
          body: JSON.stringify(tweetData),
          signal: AbortSignal.timeout(30000),
        });

        if (response.status === 201) {
          const data = await response.json() as { data?: { id?: string } };
          const tweetId = data.data?.id ?? '';
          return {
            postId: post.postId,
            platform: Platform.TWITTER,
            status: PublishStatus.PUBLISHED,
            postUrl: `https://twitter.com/i/web/status/${tweetId}`,
            retryCount: 0,
            publishedAt: new Date(),
          };
        } else {
          const errorText = await response.text();
          return {
            postId: post.postId,
            platform: Platform.TWITTER,
            status: PublishStatus.FAILED,
            errorMessage: `Twitter API error: ${response.status} - ${errorText}`,
            retryCount: 0,
          };
        }
      },
      { maxRetries: 2, baseDelayMs: 3000 }
    );
  }

  private async publishLinkedIn(post: SocialPost, _images?: ImageSet): Promise<PublishResult> {
    if (!this.config.linkedinAccessToken) {
      return {
        postId: post.postId,
        platform: Platform.LINKEDIN,
        status: PublishStatus.FAILED,
        errorMessage: 'LinkedIn credentials not configured',
        retryCount: 0,
      };
    }

    if (!(await this.rateLimiter.acquire('linkedin_post'))) {
      return {
        postId: post.postId,
        platform: Platform.LINKEDIN,
        status: PublishStatus.FAILED,
        errorMessage: 'LinkedIn rate limited',
        retryCount: 0,
      };
    }

    return retryWithBackoff(
      async () => {
        // First get user URN
        const meResponse = await fetch('https://api.linkedin.com/v2/me', {
          headers: {
            Authorization: `Bearer ${this.config.linkedinAccessToken}`,
          },
          signal: AbortSignal.timeout(30000),
        });

        if (!meResponse.ok) {
          return {
            postId: post.postId,
            platform: Platform.LINKEDIN,
            status: PublishStatus.FAILED,
            errorMessage: 'Failed to get LinkedIn user info',
            retryCount: 0,
          };
        }

        const userData = await meResponse.json() as { id?: string };
        const userUrn = `urn:li:person:${userData.id}`;

        // Create post
        const postData = {
          author: userUrn,
          lifecycleState: 'PUBLISHED',
          specificContent: {
            'com.linkedin.ugc.ShareContent': {
              shareCommentary: { text: post.refinedContent ?? post.content },
              shareMediaCategory: 'NONE',
            },
          },
          visibility: {
            'com.linkedin.ugc.MemberNetworkVisibility': 'PUBLIC',
          },
        };

        const response = await fetch('https://api.linkedin.com/v2/ugcPosts', {
          method: 'POST',
          headers: {
            Authorization: `Bearer ${this.config.linkedinAccessToken}`,
            'Content-Type': 'application/json',
            'X-Restli-Protocol-Version': '2.0.0',
          },
          body: JSON.stringify(postData),
          signal: AbortSignal.timeout(30000),
        });

        if (response.status === 201) {
          const data = await response.json() as { id?: string };
          const postUrn = data.id ?? '';
          return {
            postId: post.postId,
            platform: Platform.LINKEDIN,
            status: PublishStatus.PUBLISHED,
            postUrl: `https://www.linkedin.com/feed/update/${postUrn}`,
            retryCount: 0,
            publishedAt: new Date(),
          };
        } else {
          const errorText = await response.text();
          return {
            postId: post.postId,
            platform: Platform.LINKEDIN,
            status: PublishStatus.FAILED,
            errorMessage: `LinkedIn API error: ${response.status} - ${errorText}`,
            retryCount: 0,
          };
        }
      },
      { maxRetries: 2, baseDelayMs: 3000 }
    );
  }

  private async publishInstagram(post: SocialPost, images?: ImageSet): Promise<PublishResult> {
    if (!this.config.instagramAccessToken || !this.config.instagramBusinessId) {
      return {
        postId: post.postId,
        platform: Platform.INSTAGRAM,
        status: PublishStatus.FAILED,
        errorMessage: 'Instagram credentials not configured',
        retryCount: 0,
      };
    }

    if (!images || images.images.length === 0) {
      return {
        postId: post.postId,
        platform: Platform.INSTAGRAM,
        status: PublishStatus.FAILED,
        errorMessage: 'Instagram requires an image',
        retryCount: 0,
      };
    }

    if (!(await this.rateLimiter.acquire('instagram_post'))) {
      return {
        postId: post.postId,
        platform: Platform.INSTAGRAM,
        status: PublishStatus.FAILED,
        errorMessage: 'Instagram rate limited',
        retryCount: 0,
      };
    }

    // Carousel publishing: use Graph API CAROUSEL type when post has multiple slides
    if (post.carouselSlides && post.carouselSlides.length > 1 && images.images.length > 1) {
      return this.publishInstagramCarousel(post, images);
    }

    // Standard single-image publish
    const containerUrl = new URL(
      `https://graph.facebook.com/v18.0/${this.config.instagramBusinessId}/media`
    );
    containerUrl.searchParams.set('image_url', images.images[0].url);
    containerUrl.searchParams.set('caption', post.refinedContent ?? post.content);
    containerUrl.searchParams.set('access_token', this.config.instagramAccessToken);

    const containerResponse = await fetch(containerUrl.toString(), {
      method: 'POST',
      signal: AbortSignal.timeout(60000),
    });

    if (!containerResponse.ok) {
      const errorText = await containerResponse.text();
      return {
        postId: post.postId,
        platform: Platform.INSTAGRAM,
        status: PublishStatus.FAILED,
        errorMessage: `Instagram container error: ${errorText}`,
        retryCount: 0,
      };
    }

    const containerData = await containerResponse.json() as { id?: string };
    const containerId = containerData.id;

    const publishUrl = new URL(
      `https://graph.facebook.com/v18.0/${this.config.instagramBusinessId}/media_publish`
    );
    publishUrl.searchParams.set('creation_id', containerId ?? '');
    publishUrl.searchParams.set('access_token', this.config.instagramAccessToken);

    const publishResponse = await fetch(publishUrl.toString(), {
      method: 'POST',
      signal: AbortSignal.timeout(60000),
    });

    if (publishResponse.ok) {
      const data = await publishResponse.json() as { id?: string };
      const mediaId = data.id ?? '';
      return {
        postId: post.postId,
        platform: Platform.INSTAGRAM,
        status: PublishStatus.PUBLISHED,
        postUrl: `https://www.instagram.com/p/${mediaId}`,
        retryCount: 0,
        publishedAt: new Date(),
      };
    } else {
      const errorText = await publishResponse.text();
      return {
        postId: post.postId,
        platform: Platform.INSTAGRAM,
        status: PublishStatus.FAILED,
        errorMessage: `Instagram publish error: ${errorText}`,
        retryCount: 0,
      };
    }
  }

  /**
   * Publish an Instagram carousel using the Graph API CAROUSEL media type.
   */
  private async publishInstagramCarousel(post: SocialPost, images: ImageSet): Promise<PublishResult> {
    const businessId = this.config.instagramBusinessId!;
    const accessToken = this.config.instagramAccessToken!;
    const childContainerIds: string[] = [];

    // Step 1: Create child image containers (max 10 per carousel)
    const slideImages = images.images.slice(0, 10);
    for (const image of slideImages) {
      const childUrl = new URL(`https://graph.facebook.com/v18.0/${businessId}/media`);
      childUrl.searchParams.set('image_url', image.url);
      childUrl.searchParams.set('is_carousel_item', 'true');
      childUrl.searchParams.set('access_token', accessToken);

      const childResponse = await fetch(childUrl.toString(), {
        method: 'POST',
        signal: AbortSignal.timeout(60000),
      });

      if (!childResponse.ok) {
        const errorText = await childResponse.text();
        this.logger.warn(`Carousel child container failed: ${errorText}`);
        continue;
      }

      const childData = await childResponse.json() as { id?: string };
      if (childData.id) {
        childContainerIds.push(childData.id);
      }
    }

    if (childContainerIds.length < 2) {
      return {
        postId: post.postId,
        platform: Platform.INSTAGRAM,
        status: PublishStatus.FAILED,
        errorMessage: `Carousel requires at least 2 images, only ${childContainerIds.length} succeeded`,
        retryCount: 0,
      };
    }

    // Step 2: Create carousel container
    const carouselUrl = new URL(`https://graph.facebook.com/v18.0/${businessId}/media`);
    carouselUrl.searchParams.set('media_type', 'CAROUSEL');
    carouselUrl.searchParams.set('children', childContainerIds.join(','));
    carouselUrl.searchParams.set('caption', post.refinedContent ?? post.content);
    carouselUrl.searchParams.set('access_token', accessToken);

    const carouselResponse = await fetch(carouselUrl.toString(), {
      method: 'POST',
      signal: AbortSignal.timeout(60000),
    });

    if (!carouselResponse.ok) {
      const errorText = await carouselResponse.text();
      return {
        postId: post.postId,
        platform: Platform.INSTAGRAM,
        status: PublishStatus.FAILED,
        errorMessage: `Instagram carousel container error: ${errorText}`,
        retryCount: 0,
      };
    }

    const carouselData = await carouselResponse.json() as { id?: string };
    const carouselContainerId = carouselData.id;

    // Step 3: Publish the carousel
    const publishUrl = new URL(`https://graph.facebook.com/v18.0/${businessId}/media_publish`);
    publishUrl.searchParams.set('creation_id', carouselContainerId ?? '');
    publishUrl.searchParams.set('access_token', accessToken);

    const publishResponse = await fetch(publishUrl.toString(), {
      method: 'POST',
      signal: AbortSignal.timeout(60000),
    });

    if (publishResponse.ok) {
      const data = await publishResponse.json() as { id?: string };
      const mediaId = data.id ?? '';
      this.logger.info(`Instagram carousel published: ${childContainerIds.length} slides`);
      return {
        postId: post.postId,
        platform: Platform.INSTAGRAM,
        status: PublishStatus.PUBLISHED,
        postUrl: `https://www.instagram.com/p/${mediaId}`,
        retryCount: 0,
        publishedAt: new Date(),
      };
    } else {
      const errorText = await publishResponse.text();
      return {
        postId: post.postId,
        platform: Platform.INSTAGRAM,
        status: PublishStatus.FAILED,
        errorMessage: `Instagram carousel publish error: ${errorText}`,
        retryCount: 0,
      };
    }
  }

  private async publishFacebook(post: SocialPost, images?: ImageSet): Promise<PublishResult> {
    if (!this.config.facebookAccessToken || !this.config.facebookPageId) {
      return {
        postId: post.postId,
        platform: Platform.FACEBOOK,
        status: PublishStatus.FAILED,
        errorMessage: 'Facebook credentials not configured',
        retryCount: 0,
      };
    }

    if (!(await this.rateLimiter.acquire('facebook_post'))) {
      return {
        postId: post.postId,
        platform: Platform.FACEBOOK,
        status: PublishStatus.FAILED,
        errorMessage: 'Facebook rate limited',
        retryCount: 0,
      };
    }

    const formData = new URLSearchParams();
    formData.append('message', post.refinedContent ?? post.content);
    formData.append('access_token', this.config.facebookAccessToken);

    if (images && images.images.length > 0) {
      formData.append('link', images.images[0].url);
    }

    const response = await fetch(
      `https://graph.facebook.com/v18.0/${this.config.facebookPageId}/feed`,
      {
        method: 'POST',
        body: formData,
        signal: AbortSignal.timeout(30000),
      }
    );

    if (response.ok) {
      const data = await response.json() as { id?: string };
      const fbPostId = data.id ?? '';
      return {
        postId: post.postId,
        platform: Platform.FACEBOOK,
        status: PublishStatus.PUBLISHED,
        postUrl: `https://www.facebook.com/${fbPostId}`,
        retryCount: 0,
        publishedAt: new Date(),
      };
    } else {
      const errorText = await response.text();
      return {
        postId: post.postId,
        platform: Platform.FACEBOOK,
        status: PublishStatus.FAILED,
        errorMessage: `Facebook API error: ${response.status} - ${errorText}`,
        retryCount: 0,
      };
    }
  }

  private async publishTikTok(post: SocialPost, _images?: ImageSet): Promise<PublishResult> {
    // TikTok's Content Posting API has limited availability
    return {
      postId: post.postId,
      platform: Platform.TIKTOK,
      status: PublishStatus.PENDING_REVIEW,
      errorMessage: 'TikTok posting requires manual upload or approved API access',
      retryCount: 0,
    };
  }

  private simulatePublish(post: SocialPost, images?: ImageSet): PublishResult {
    this.logger.info(`[DRY RUN] Would publish to ${post.platform}:`);
    this.logger.info(`  Content: ${(post.refinedContent ?? post.content).slice(0, 100)}...`);
    if (images && images.images.length > 0) {
      this.logger.info(`  Images: ${images.images.length}`);
    }

    return {
      postId: post.postId,
      platform: post.platform,
      status: PublishStatus.PUBLISHED,
      postUrl: `https://example.com/dry-run/${post.postId}`,
      retryCount: 0,
      publishedAt: new Date(),
    };
  }

  getManualReviewQueue(): PublishResult[] {
    return [...this.manualReviewQueue];
  }

  clearManualReviewQueue(): void {
    this.manualReviewQueue = [];
  }

  /**
   * Build an OAuth 1.0a Authorization header for Twitter API v2.
   */
  private buildOAuth1Header(method: string, url: string): string {
    const consumerKey = this.config.twitterApiKey!;
    const consumerSecret = this.config.twitterApiSecret!;
    const accessToken = this.config.twitterAccessToken!;
    const accessSecret = this.config.twitterAccessSecret!;

    const oauthParams: Record<string, string> = {
      oauth_consumer_key: consumerKey,
      oauth_nonce: crypto.randomBytes(16).toString('hex'),
      oauth_signature_method: 'HMAC-SHA1',
      oauth_timestamp: Math.floor(Date.now() / 1000).toString(),
      oauth_token: accessToken,
      oauth_version: '1.0',
    };

    // Build signature base string
    const paramString = Object.keys(oauthParams)
      .sort()
      .map((k) => `${this.percentEncode(k)}=${this.percentEncode(oauthParams[k])}`)
      .join('&');

    const baseString = [
      method.toUpperCase(),
      this.percentEncode(url),
      this.percentEncode(paramString),
    ].join('&');

    const signingKey = `${this.percentEncode(consumerSecret)}&${this.percentEncode(accessSecret)}`;
    const signature = crypto
      .createHmac('sha1', signingKey)
      .update(baseString)
      .digest('base64');

    oauthParams['oauth_signature'] = signature;

    const header = Object.keys(oauthParams)
      .sort()
      .map((k) => `${this.percentEncode(k)}="${this.percentEncode(oauthParams[k])}"`)
      .join(', ');

    return `OAuth ${header}`;
  }

  private percentEncode(str: string): string {
    return encodeURIComponent(str).replace(/[!'()*]/g, (c) => `%${c.charCodeAt(0).toString(16).toUpperCase()}`);
  }
}

File: orchestrator/agents/publish_agent.py
"""Publishing Agent - Posts content to social media platforms."""

import asyncio
from datetime import datetime
from typing import Dict, List, Optional

import aiohttp

from .base_agent import BaseAgent
from ..models import (
    SocialPost,
    ImageSet,
    PublishResult,
    PublishStatus,
    Platform,
)
from ..utils import retry_with_backoff, RateLimiter, get_logger
from ..config import Config

logger = get_logger(__name__)


class PublishAgent(BaseAgent):
    """
    Agent #4: Publishing
    
    Posts content to multiple social media platforms:
    - Twitter/X
    - LinkedIn
    - Instagram
    - Facebook
    - TikTok
    
    Handles OAuth authentication, retry logic, and status tracking.
    """
    
    def __init__(self, config: Config, rate_limiter: Optional[RateLimiter] = None):
        super().__init__("publish_agent", rate_limiter)
        self.config = config
        self._max_retries = 3
        self._manual_review_queue: List[PublishResult] = []
    
    async def execute(
        self,
        posts: List[SocialPost],
        image_sets: List[ImageSet],
        dry_run: bool = False,
    ) -> List[PublishResult]:
        """
        Publish posts to their target platforms.
        
        Args:
            posts: List of posts to publish
            image_sets: List of image sets (matched by post_id)
            dry_run: If True, simulate publishing without actually posting
            
        Returns:
            List of PublishResult objects with status for each post
        """
        self.logger.info(
            f"Publishing {len(posts)} posts (dry_run={dry_run})"
        )
        
        # Create image lookup
        images_by_post: Dict[str, ImageSet] = {
            img_set.post_id: img_set for img_set in image_sets
        }
        
        results = []
        
        for post in posts:
            images = images_by_post.get(post.post_id)
            
            if dry_run:
                result = self._simulate_publish(post, images)
            else:
                result = await self._publish_post(post, images)
            
            results.append(result)
            
            # Stagger posts to avoid spam appearance
            if not dry_run:
                await asyncio.sleep(2)  # 2 second delay between posts
        
        # Log summary
        success_count = sum(1 for r in results if r.status == PublishStatus.PUBLISHED)
        failed_count = sum(1 for r in results if r.status == PublishStatus.FAILED)
        review_count = sum(1 for r in results if r.status == PublishStatus.PENDING_REVIEW)
        
        self.logger.info(
            f"Publishing complete: {success_count} success, "
            f"{failed_count} failed, {review_count} pending review"
        )
        
        return results
    
    async def _publish_post(
        self,
        post: SocialPost,
        images: Optional[ImageSet],
    ) -> PublishResult:
        """Publish a single post with retry logic."""
        platform_handlers = {
            Platform.TWITTER: self._publish_twitter,
            Platform.LINKEDIN: self._publish_linkedin,
            Platform.INSTAGRAM: self._publish_instagram,
            Platform.FACEBOOK: self._publish_facebook,
            Platform.TIKTOK: self._publish_tiktok,
        }
        
        handler = platform_handlers.get(post.platform)
        if not handler:
            return PublishResult(
                post_id=post.post_id,
                platform=post.platform,
                status=PublishStatus.FAILED,
                error_message=f"Unsupported platform: {post.platform}",
            )
        
        last_error = None
        for attempt in range(self._max_retries):
            try:
                result = await handler(post, images)
                if result.status == PublishStatus.PUBLISHED:
                    return result
                last_error = result.error_message
            except Exception as e:
                last_error = str(e)
                self.logger.warning(
                    f"Publish attempt {attempt + 1} failed for {post.post_id}: {e}"
                )
                
                if attempt < self._max_retries - 1:
                    await asyncio.sleep(5 * (attempt + 1))  # Exponential backoff
        
        # All retries failed - queue for manual review
        result = PublishResult(
            post_id=post.post_id,
            platform=post.platform,
            status=PublishStatus.PENDING_REVIEW,
            error_message=f"Max retries exceeded: {last_error}",
            retry_count=self._max_retries,
        )
        self._manual_review_queue.append(result)
        return result
    
    @retry_with_backoff(max_retries=2, base_delay=3.0)
    async def _publish_twitter(
        self,
        post: SocialPost,
        images: Optional[ImageSet],
    ) -> PublishResult:
        """Publish to Twitter/X using API v2."""
        if not self.config.twitter_access_token:
            return PublishResult(
                post_id=post.post_id,
                platform=Platform.TWITTER,
                status=PublishStatus.FAILED,
                error_message="Twitter credentials not configured",
            )
        
        if not await self.rate_limiter.acquire("twitter_post"):
            return PublishResult(
                post_id=post.post_id,
                platform=Platform.TWITTER,
                status=PublishStatus.FAILED,
                error_message="Twitter rate limited",
            )
        
        # First upload media if present
        media_ids = []
        if images and images.images:
            for image in images.images[:4]:  # Twitter allows max 4 images
                try:
                    media_id = await self._upload_twitter_media(str(image.url))
                    if media_id:
                        media_ids.append(media_id)
                except Exception as e:
                    self.logger.warning(f"Failed to upload media: {e}")
        
        async with aiohttp.ClientSession() as session:
            tweet_data = {"text": post.content}
            if media_ids:
                tweet_data["media"] = {"media_ids": media_ids}
            
            async with session.post(
                "https://api.twitter.com/2/tweets",
                headers={
                    "Authorization": f"Bearer {self.config.twitter_access_token}",
                    "Content-Type": "application/json",
                },
                json=tweet_data,
                timeout=30,
            ) as response:
                if response.status == 201:
                    data = await response.json()
                    tweet_id = data["data"]["id"]
                    return PublishResult(
                        post_id=post.post_id,
                        platform=Platform.TWITTER,
                        status=PublishStatus.PUBLISHED,
                        post_url=f"https://twitter.com/i/web/status/{tweet_id}",
                        published_at=datetime.utcnow(),
                    )
                else:
                    error_text = await response.text()
                    return PublishResult(
                        post_id=post.post_id,
                        platform=Platform.TWITTER,
                        status=PublishStatus.FAILED,
                        error_message=f"Twitter API error: {response.status} - {error_text}",
                    )
    
    async def _upload_twitter_media(self, image_url: str) -> Optional[str]:
        """Upload media to Twitter for attachment."""
        # Twitter media upload requires downloading and re-uploading
        # This is a simplified placeholder
        return None
    
    @retry_with_backoff(max_retries=2, base_delay=3.0)
    async def _publish_linkedin(
        self,
        post: SocialPost,
        images: Optional[ImageSet],
    ) -> PublishResult:
        """Publish to LinkedIn using API."""
        if not self.config.linkedin_access_token:
            return PublishResult(
                post_id=post.post_id,
                platform=Platform.LINKEDIN,
                status=PublishStatus.FAILED,
                error_message="LinkedIn credentials not configured",
            )
        
        if not await self.rate_limiter.acquire("linkedin_post"):
            return PublishResult(
                post_id=post.post_id,
                platform=Platform.LINKEDIN,
                status=PublishStatus.FAILED,
                error_message="LinkedIn rate limited",
            )
        
        async with aiohttp.ClientSession() as session:
            # First get user URN
            async with session.get(
                "https://api.linkedin.com/v2/me",
                headers={
                    "Authorization": f"Bearer {self.config.linkedin_access_token}",
                },
            ) as response:
                if response.status != 200:
                    return PublishResult(
                        post_id=post.post_id,
                        platform=Platform.LINKEDIN,
                        status=PublishStatus.FAILED,
                        error_message="Failed to get LinkedIn user info",
                    )
                user_data = await response.json()
                user_urn = f"urn:li:person:{user_data['id']}"
            
            # Create post
            post_data = {
                "author": user_urn,
                "lifecycleState": "PUBLISHED",
                "specificContent": {
                    "com.linkedin.ugc.ShareContent": {
                        "shareCommentary": {"text": post.content},
                        "shareMediaCategory": "NONE",
                    }
                },
                "visibility": {
                    "com.linkedin.ugc.MemberNetworkVisibility": "PUBLIC"
                },
            }
            
            async with session.post(
                "https://api.linkedin.com/v2/ugcPosts",
                headers={
                    "Authorization": f"Bearer {self.config.linkedin_access_token}",
                    "Content-Type": "application/json",
                    "X-Restli-Protocol-Version": "2.0.0",
                },
                json=post_data,
                timeout=30,
            ) as response:
                if response.status == 201:
                    data = await response.json()
                    post_urn = data.get("id", "")
                    return PublishResult(
                        post_id=post.post_id,
                        platform=Platform.LINKEDIN,
                        status=PublishStatus.PUBLISHED,
                        post_url=f"https://www.linkedin.com/feed/update/{post_urn}",
                        published_at=datetime.utcnow(),
                    )
                else:
                    error_text = await response.text()
                    return PublishResult(
                        post_id=post.post_id,
                        platform=Platform.LINKEDIN,
                        status=PublishStatus.FAILED,
                        error_message=f"LinkedIn API error: {response.status} - {error_text}",
                    )
    
    async def _publish_instagram(
        self,
        post: SocialPost,
        images: Optional[ImageSet],
    ) -> PublishResult:
        """Publish to Instagram using Graph API."""
        if not self.config.instagram_access_token or not self.config.instagram_business_id:
            return PublishResult(
                post_id=post.post_id,
                platform=Platform.INSTAGRAM,
                status=PublishStatus.FAILED,
                error_message="Instagram credentials not configured",
            )
        
        if not images or not images.images:
            return PublishResult(
                post_id=post.post_id,
                platform=Platform.INSTAGRAM,
                status=PublishStatus.FAILED,
                error_message="Instagram requires an image",
            )
        
        if not await self.rate_limiter.acquire("instagram_post"):
            return PublishResult(
                post_id=post.post_id,
                platform=Platform.INSTAGRAM,
                status=PublishStatus.FAILED,
                error_message="Instagram rate limited",
            )
        
        async with aiohttp.ClientSession() as session:
            # Step 1: Create media container
            async with session.post(
                f"https://graph.facebook.com/v18.0/{self.config.instagram_business_id}/media",
                params={
                    "image_url": str(images.images[0].url),
                    "caption": post.content,
                    "access_token": self.config.instagram_access_token,
                },
                timeout=60,
            ) as response:
                if response.status != 200:
                    error_text = await response.text()
                    return PublishResult(
                        post_id=post.post_id,
                        platform=Platform.INSTAGRAM,
                        status=PublishStatus.FAILED,
                        error_message=f"Instagram container error: {error_text}",
                    )
                data = await response.json()
                container_id = data["id"]
            
            # Step 2: Publish the container
            async with session.post(
                f"https://graph.facebook.com/v18.0/{self.config.instagram_business_id}/media_publish",
                params={
                    "creation_id": container_id,
                    "access_token": self.config.instagram_access_token,
                },
                timeout=60,
            ) as response:
                if response.status == 200:
                    data = await response.json()
                    media_id = data["id"]
                    return PublishResult(
                        post_id=post.post_id,
                        platform=Platform.INSTAGRAM,
                        status=PublishStatus.PUBLISHED,
                        post_url=f"https://www.instagram.com/p/{media_id}",
                        published_at=datetime.utcnow(),
                    )
                else:
                    error_text = await response.text()
                    return PublishResult(
                        post_id=post.post_id,
                        platform=Platform.INSTAGRAM,
                        status=PublishStatus.FAILED,
                        error_message=f"Instagram publish error: {error_text}",
                    )
    
    async def _publish_facebook(
        self,
        post: SocialPost,
        images: Optional[ImageSet],
    ) -> PublishResult:
        """Publish to Facebook Page using Graph API."""
        if not self.config.facebook_access_token or not self.config.facebook_page_id:
            return PublishResult(
                post_id=post.post_id,
                platform=Platform.FACEBOOK,
                status=PublishStatus.FAILED,
                error_message="Facebook credentials not configured",
            )
        
        if not await self.rate_limiter.acquire("facebook_post"):
            return PublishResult(
                post_id=post.post_id,
                platform=Platform.FACEBOOK,
                status=PublishStatus.FAILED,
                error_message="Facebook rate limited",
            )
        
        async with aiohttp.ClientSession() as session:
            post_data = {
                "message": post.content,
                "access_token": self.config.facebook_access_token,
            }
            
            # Add image if present
            if images and images.images:
                post_data["link"] = str(images.images[0].url)
            
            async with session.post(
                f"https://graph.facebook.com/v18.0/{self.config.facebook_page_id}/feed",
                data=post_data,
                timeout=30,
            ) as response:
                if response.status == 200:
                    data = await response.json()
                    fb_post_id = data["id"]
                    return PublishResult(
                        post_id=post.post_id,
                        platform=Platform.FACEBOOK,
                        status=PublishStatus.PUBLISHED,
                        post_url=f"https://www.facebook.com/{fb_post_id}",
                        published_at=datetime.utcnow(),
                    )
                else:
                    error_text = await response.text()
                    return PublishResult(
                        post_id=post.post_id,
                        platform=Platform.FACEBOOK,
                        status=PublishStatus.FAILED,
                        error_message=f"Facebook API error: {response.status} - {error_text}",
                    )
    
    async def _publish_tiktok(
        self,
        post: SocialPost,
        images: Optional[ImageSet],
    ) -> PublishResult:
        """Publish to TikTok (placeholder - TikTok API is limited for posting)."""
        # Note: TikTok's Content Posting API has limited availability
        # This is a placeholder implementation
        return PublishResult(
            post_id=post.post_id,
            platform=Platform.TIKTOK,
            status=PublishStatus.PENDING_REVIEW,
            error_message="TikTok posting requires manual upload or approved API access",
        )
    
    def _simulate_publish(
        self,
        post: SocialPost,
        images: Optional[ImageSet],
    ) -> PublishResult:
        """Simulate publishing for dry-run mode."""
        self.logger.info(f"[DRY RUN] Would publish to {post.platform.value}:")
        self.logger.info(f"  Content: {post.content[:100]}...")
        if images and images.images:
            self.logger.info(f"  Images: {len(images.images)}")
        
        return PublishResult(
            post_id=post.post_id,
            platform=post.platform,
            status=PublishStatus.PUBLISHED,
            post_url=f"https://example.com/dry-run/{post.post_id}",
            published_at=datetime.utcnow(),
        )
    
    def get_manual_review_queue(self) -> List[PublishResult]:
        """Get posts queued for manual review."""
        return self._manual_review_queue.copy()
    
    def clear_manual_review_queue(self):
        """Clear the manual review queue."""
        self._manual_review_queue.clear()
